### PR TITLE
Use systemd socket directly instead of libsystemd

### PR DESCRIPTION
**Pull request by rjernst**
The libsystemd library function sd_notify is just a thin wrapper around opeing and writing to a unix filesystem socket. This commit replaces using libsystemd with opening the socket provided by systemd directly.

relates #86475

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Do we need JNA at all here? AFAIK all the JDKs we care about include support for unix domain sockets.

**Comment by rjernst** [comment]
Unfortunately the unix sockets added to Java are not enough. I tried that at first, and when it didn't work, found the following:
https://bugs.openjdk.org/browse/JDK-8297837

tldr; datagram sockets aren't supported, which we need for systemd

**Comment by DaveCTurner** [comment]
> tldr; datagram sockets aren't supported, which we need for systemd

Ugh ok I see, thanks for clarifying. Could you mention that in a code comment?

**Review by ldematte — APPROVED** [review]
LGTM, just a couple of questions

File: libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkPosixCLibrary.java [hunk 1]
```diff
@@ -326,4 +379,15 @@ public long bytesalloc() {
...
     }
+
+    private static class JdkSockAddr implements SockAddr {
+        private static final MemoryLayout layout = MemoryLayout.structLayout(JAVA_SHORT, MemoryLayout.sequenceLayout(108, JAVA_BYTE));
+        final MemorySegment segment;
+
+        JdkSockAddr(String path) {
+            segment = Arena.ofAuto().allocate(layout);
+            segment.set(JAVA_SHORT, 0, AF_UNIX);
+            MemorySegmentUtil.setString(segment, 2, path);
```
[the remark this answers is not in the document]
Comment (rjernst) [reply libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkPosixCLibrary.java:390]: You're right it won't matter (system only has messages in ascii), but I'll bring them in line with each other.

File: libs/native/src/main/java/org/elasticsearch/nativeaccess/Systemd.java [hunk 2]
```diff
@@ -41,15 +45,56 @@ public void notify_stopping() {
...
+            return;
+        }
+        RuntimeException error = null;
+        try {
+            var sockAddr = libc.newUnixSockAddr(socketPath);
+            if (libc.connect(sockfd, sockAddr) != 0) {
+                throwOrLog("Could not connect to systemd socket: " + libc.strerror(libc.errno()), warnOnError);
+                return;
+            }
+            buffer.buffer().clear();
```
[the remark this answers is not in the document]
Comment (rjernst) [reply libs/native/src/main/java/org/elasticsearch/nativeaccess/Systemd.java:60]: I'll look into this. I thought the existing solution already wasn't threadsafe, but i see that we do possibly call these from different threads in SystemdPlugin

[the remark this answers is not in the document]
Comment (rjernst) [reply libs/native/src/main/java/org/elasticsearch/nativeaccess/Systemd.java:60]: Since this isn't performance sensitive (we only do a couple of these calls, usually not near each other), I just add a simple synchronized block on the buffer around its use. Happy to change that further.
